criu/util.c

```diff
@@ -308,10 +311,18 @@ static inline int set_proc_pid_fd(int pid, int fd)
 static inline int get_proc_fd(int pid)
 {
 	if (pid == PROC_SELF) {
-		if (open_proc_self_fd != -1 && open_proc_self_pid != getpid()) {
-			close(open_proc_self_fd);
+		int open_proc_self_fd;
+
+		open_proc_self_fd = get_service_fd(PROC_SELF_FD_OFF);
+		/**
+		 * FIXME in case two processes from different pidnses share
+		 * both files and pids - this check will not detect that we
+		 * reuse /proc/self of a different process and everything proc
+		 * related will break. Note that service_fd_id does not help
+		 * here either.
```

Comment:
if you will close PROC_SELF_FD_OFF in clone_service_fd, you will not have this problem.

Reply:
Function clone_service_fd() is called from:
```
  +-< clone_service_fd
    +-< setup_newborn_fds
    | +-< restore_task_with_children
```
So for restored tasks/helpers it helps. But there are different call_in_child_process() or just fork() kind of processes for which this problem is probably still there.